Adds configurable sticky columns to KTable

marcellamaki opened the pull request:
<!-- Please remove any unused sections -->

## Description
This PR allows for configurable stickiness for the first, first two, and last columns on a table

#### Issue addressed
<!-- Only necessary if applicable -->

Addresses [#1005](https://github.com/learningequality/kolibri-design-system/issues/1005)

### Before/after screenshots

default table with no content that overflows width; default table (i.e. no changes made by me to the documentation page for this example) when the content overflows the width; specific sticky column implementation

https://github.com/user-attachments/assets/6d9dbf95-6554-4efd-8be5-d849d92d135e



## Changelog
<!-- [DO NOT REMOVE-USED BY GH ACTION] CHANGELOG START -->

<!--
  - Fill in the changelog item(s) below. If there are more groups of closely
    related changes, prepare more changelog items for each one of them.
    At a minimum, always separete non-breaking changes from breaking changes.
  - This needs to be pasted to CHANGELOG.md before merging a PR.
  - See changelog guidelines https://www.notion.so/learningequality/DRAFT-Changelog-Guidelines-106b6ebbdeda4ba5b3b3e7c490c5a4fe and existing
    items in CHANGELOG.md as examples
 -->

  - **Description:** This PR adds optional sticky columns for the first, firstTwo, and last columns. The default is 'first' only if nothing else is specified, and the stickiness is only apparent is the contents of the table is wider than the container
  - **Products impact:** Choose from - none (for internal updates) / bugfix / new API / updated API / removed API. If it's 'none', use "-" for all items below to indicate they are not relevant.
  - **Addresses:** [Link(s) to GH issue(s) addressed. Include KDS links as well as links to related issues in a consumer product repository too.](https://github.com/learningequality/kolibri-design-system/issues/1005). Part of 0.19 Kolibri work
  - **Components:** KTable
  - **Breaking:** No
  - **Impacts a11y:** No
  - **Guidance:** The "default" behavior of the first column being sticky will be introduced when KDS is updated after this is merged. Use of the additional options depends on the content of the table, and may need to be customized for mobile responsiveness

<!-- [DO NOT REMOVE-USED BY GH ACTION] CHANGELOG END -->

## Steps to test

1. Preview the additions to the KTable docs page built in the PR, specifically the sticky table example

## (optional) Implementation notes

### At a high level, how did you implement this?
<!-- Briefly describe how this works -->

### Does this introduce any tech-debt items?
No 

## Testing checklist
<!-- Complete the checklist before submitting a PR; delete anything that doesn't apply -->

- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical and brittle code paths are covered by unit tests
- [ ] The change is described in the changelog section above

## Reviewer guidance
<!-- Delete anything that doesn't apply so your reviewer knows what to check for -->

- [ ] Is the code clean and well-commented?
- [ ] Are there tests for this change?
- [ ] Are all UI components LTR and RTL compliant (if applicable)?
- [ ] _Add other things to check for here_

## Comments
Is the first column being sticky a reasonable default? 

I considered a more expansive API for additional column stickiness, but practically speaking, this is the only scenario we need. Since we are also having a broader conversation about some KTable refactors for a more dev friendly experience and reducing the use of column indices, etc. I decided to start with the most narrowly scoped approach to solve the problem at hand. I also can quite think of how a middle-of-the-table column stickiness would help anything. I also considered the value of using the indices vs. using a string prop with validation. I find the user experience of the 'first', 'firstTwo', 'last' to be easiest to follow, but would be interested in any other perspectives here.

marcellamaki committed:
Update KTable to include configurable sticky columns for many-column wide tables and mobile responsiveness

LianaHarris360 commented on lib/KTable/index.vue:
```diff
@@ -718,24 +856,46 @@
     position: sticky;
     top: 0;
-    z-index: 2;
+    z-index: 10;
```
It looks like the increased z-index values are causing the headers and columns to be displayed over the AppBar in Kolibri, because the [ScrollingHeader](https://github.com/learningequality/kolibri/blob/6b54f83bade08d2a989df9f525457422b59ccba1/packages/kolibri/components/pages/ScrollingHeader.vue#L166) has a z-index of 4.

![Screenshot 2025-06-26](https://github.com/user-attachments/assets/c21ba994-dde4-4978-913d-31c2e3267726)

marcellamaki replied:
ooo yikes! good catch - I'll update this

LianaHarris360 commented on lib/KTable/index.vue:
```diff
@@ -182,6 +192,12 @@
 
       const isTableEmpty = computed(() => sortedRows.value.length === 0);
 
+      const isTableScrollable = computed(() => {
```
On the initial load, this value is false and the vertical lines beside the sticky columns don't appear, even though the table can be scrolled.

![Screenshot 2025-06AM](https://github.com/user-attachments/assets/872e0d90-2538-4fdd-afe7-1c59c76b79dd)

marcellamaki replied:
The woes of testing things locally over and over! :D thanks

LianaHarris360 reviewed (dismissed):
I agree that using 'first', 'firstTwo', and 'last' is the easiest user experience to follow and to validate against. I ran these changes in Kolibri and only noted two things, otherwise the rest of the code makes sense to me.

marcellamaki committed:
Add documentation pages

MisRob commented on lib/KTable/index.vue:
```diff
@@ -182,6 +192,12 @@
 
       const isTableEmpty = computed(() => sortedRows.value.length === 0);
 
+      const isTableScrollable = computed(() => {
```
I think we could optimize this method for performance. If I recall, `useKResponsiveElement` is already optimized well so that should be covered. For other logic, I can think of

- utilizing Vue `ref` should be more performant `querySelector`
- maybe we should debounce or even better `requestAnimationFrame` for reading `table.scrollWidth`?

@AlexVelezLl did quite detailed performance review of `KTable` some time ago. How does this sound or would you recommend something else? Also I see more places in code that read from DOM elements. Except this place, I haven't yet noticed anything obviously suspicious - if you'd like to check on it too, that'd be welcome.

AlexVelezLl replied:
I agree, `useKResponsiveElement` doesn't do any kind of throttling nor debouncing when setting the sizes values so this computed property will be being recalculated a lot of times per second if the user resizes its window . I think throttling would be best here rather than debouncing to have a smoother UX if the user is resizing the window. 

I wonder @MisRob if would it worth to have a `throttle` prop for `useKResponsiveElement` that receives the number of ms to pass to the throttle function and wrap this [updateEl](https://github.com/marcellamaki/kolibri-design-system/blob/two-sticky-columns/lib/composables/useKResponsiveElement.js#L32) with a throttled function? I think throttling the watcher handler of the watch that observes the result of the `useKResposiveElement` is starting to be common in some places where we use this composable.

MisRob replied:
@AlexVelezLl I think that `useKResponsiveElement` is optimized pretty well [here](https://github.com/learningequality/kolibri-design-system/blob/develop/lib/composables/useKResponsiveElement.js#L6-L14)? Since it uses (1) ResizeObserver and (2) requestAnimationFrame. I was more concerned with other parts such as `table.scrollWidth` and `querySelector`.

AlexVelezLl replied:
Kindof, yes, but we still have so many recalculations that are perhaps not that necessary as we dont need that high reactivity. Just added a console.log in the computed property:

https://github.com/user-attachments/assets/3d7ff008-f19b-4a9a-821c-bfdcbbce5f4c

I think that's why, for example, in [KBreacrumbs](https://github.com/marcellamaki/kolibri-design-system/blob/1a370c62ebab0f838077018d5828b97366957499/lib/KBreadcrumbs.vue#L240) and [KListWithOverflow](https://github.com/marcellamaki/kolibri-design-system/blob/1a370c62ebab0f838077018d5828b97366957499/lib/KListWithOverflow.vue#L104) we throttle the watch handler function on top of the ResizeObserver and requestAnimationFrame that the useKResponsiveElement already provides.

And yes, the querySelector is the most concerning part here. I did a quick profile resizing the page, and in a 4s session, the scripting time was taking more than 1s, which is a lot. Not sure if reading the `table.scrollWidth` here would cause reflows, as this calculation is being made inside a computed property function, I would think that Vue aligns these calculations well in its render cycles, but not completely sure :sweat_smile:.

AlexVelezLl replied:
I just realized that this is just to set the shadows, we wouldnt even need a throttle then, I dont think we need that level of reactivity for this :sweat_smile:, just wrapping this in a debounce and make this recalculation just once per resizing session will mitigate almost all of the performance impact that the calculation function would cause, as it will just be executed once.

marcellamaki replied:
great thank you! I'll do that :)

MisRob replied:
Ah yes that makes sense. Thanks for the measurement! And on top of that I think we can actually remove `querySelector` in favor of `ref`. Combining with debounce should then hopefully result in good performance overall.

I think this will help not only with resizing browser window but also anytime we change table width dynamically - e.g. when changing screen orientation on tablets or phones, where I would guess performance impact may be more apparent.

AlexVelezLl replied:
Yes, I agree!

MisRob commented on lib/KTable/index.vue:
```diff
@@ -685,6 +744,85 @@
+        const stickyIndices = this.getStickyColumns();
+        return stickyIndices.includes(colIndex);
+      },
+      getStickyColumnStyle(rowIndex, colIndex) {
```
Looking at how `KTable` file grows, moving sticky logic like this, and related, to a composable would help.

MisRob reviewed (commented):
This looks very nice overall. Regarding API, within what we're able to achieve now, this feels simple and intuitive. In code, haven't dived into reviewing all calculations in detail, rather just brief skim.

I think this is a visual regression - on the latest `develop`, rows are higher than in this PR's preview.
| develop | PR |
| ----------- | ----- |
| ![01-before](https://github.com/user-attachments/assets/96c8f404-2843-40a7-874e-3bebb1168876) | ![01-after](https://github.com/user-attachments/assets/db699045-6868-45a1-9388-9937786ebba7) |

Seems to be the same issue that Percy has captured. 

I don't think all examples are covered with automated visual tests yet though, so would be good to compare [before](https://design-system.learningequality.org/ktable) / [after](https://deploy-preview-1056--kolibri-design-system.netlify.app/ktable) live documentation examples.

Relatedly, it'd be good to cover new sticky UI with visual tests.

MisRob commented:
Also, in the recording I see you added a nice new section to the documentation - seems you've just haven't yet commit that file?

By the way, after that, you could easily re-use the very same documentation example for a visual test. We've recently done the same for `KImg`. You can see [here](https://github.com/learningequality/kolibri-design-system/pull/1038/files#diff-ada10de317adf2ebcf0259662f30a85eccfe62bea0cf7197da57a0fa271e272bR22) and [here](https://github.com/learningequality/kolibri-design-system/pull/1038/files#diff-207a6b2bab9de60ed782444545a36bdc162ca91f8475fb123eeebe24db128ebcR104-R107). I think that two snapshots of this documentation example - one on mobile, and one on larger screen, would be sufficient.

marcellamaki committed:
Add visual tests for KTable

LianaHarris360 commented on lib/KTable/index.vue:
```diff
@@ -207,6 +227,12 @@
 
+      onMounted(() => {
+        nextTick(() => {
+          debounce(checkScrollable, 100);
```
The z-index issues have been resolved! I still see the `isTableScrollable` value as false on scrollable tables. After some investigation I realized that this debounced function needs to be assigned to a variable and then called here for it to eventually run `checkScrollable()`.

```
      const debouncedCheck = debounce(checkScrollable, 100);

      onMounted(() => {
        nextTick(() => {
          debouncedCheck();
        });
      });

```

marcellamaki commented:
Very belated, I have pushed my changes @LianaHarris360 @AlexVelezLl @MisRob. 

Particular focus areas/feedback requests for review: 

1. Have I actually addressed all of your feedback? Did I misunderstand any of your suggestions/is there anything I should further refine or fix?
2. I feel the least confidence about the visual tests. They pass, but I wrote them with the help of Claude. I did review them in comparison to the other visual tests we have but.... are they useful? Are they "right enough" for now? Are there important cases I'm missing? (an example of why I feel not so confident: in my Pr i accidentally deleted the overflow: auto behavior and the tests still passed, where that is clearly a visual regression. so, I may need to add more cases. does that seem worth doing now? in follow up?)
3. Any other feedback about the approach of the API? 
4. @MisRob regarding: 
> I think this is a visual regression - on the latest develop, rows are higher than in this PR's preview.

I have updated this to set a min-height for the row height, but looking at Tomiwa's most recent spec for the table in 0.19, his rows are a bit less "roomy" than the previous KTable rows, and the height here in this PR is more aligned with that (~40px) vs what it was previously (~60px). Does this feel like an appropriate change, or would you prefer I don't make any changes to the height at all. It isn't really in scope of this PR

LianaHarris360 commented on lib/KTable/KTableGridItem.vue:
```diff
@@ -89,6 +91,9 @@
 <style scoped>
 
   .cell-content {
+    display: flex;
```
I don't see it happening in Kolibri, but compared to the latest develop, the numbers for the `Age` column are no longer right-aligned.

![KTable](https://github.com/user-attachments/assets/be32fdef-7f29-49dc-984b-b8df1ddd6b20)

LianaHarris360 commented on lib/KTable/index.vue:
```diff
@@ -207,6 +227,12 @@
 
+      onMounted(() => {
+        nextTick(() => {
+          debounce(checkScrollable, 300)();
```
Since `checkScrollable` is only called during the `onMounted` hook, if the table isn’t initially scrollable and is later resized to become scrollable, `isTableScrollable` doesn't gets set to true and the horizontal lines don’t appear. 

I wonder if, rather than setting the horizontal lines and table stickiness solely based on whether the table is scrollable, we should consider automatically applying the table stickiness and displaying the lines regardless of the current scrollability. Or, if it would be better to just watch the `elementWidth` and call `checkScrollable` whenever it changes.

LianaHarris360 reviewed (commented):
I think it's appropriate to set the min-height within this PR, since it aligns with our expectations for how KTable will eventually look. I also agree with Misha that moving the sticky logic into a composable could be beneficial, but that can be addressed separately, it doesn’t necessarily need to be included in this PR.

MisRob commented:
Hi @marcellamaki, thanks for all the updates.

> I’ve set a min-height for the row, but Tomiwa’s latest table spec for 0.19 uses less “roomy” rows than before.

Sounds like this is an intentional change for all `KTable`s. No problem at all then.

---

Regarding visual tests, my original suggestion was just to import `docs/examples/KTable/StickyColumns.vue` into the visual tests, rather than duplicate the example. KDS lets us share examples between docs and visual tests. I see you’ve covered more states—thank you! Just a heads-up: we’re preparing a refactor for visual tests, and the process will change a bit ([see guide](https://github.com/learningequality/kolibri-design-system/blob/3d7bad3933dcb7d45425fd9bd6f3c4a5ca048e14/dev_docs/07_visual_testing_guide.md)). We can still use work from this PR, but there will be some conflicts with https://github.com/learningequality/kolibri-design-system/pull/1107. Let’s chat about how to proceed—I’m sure we can resolve it without blocking this PR.

---

> Have I actually addressed all of your feedback?

Not sure if you saw my note about moving sticky logic to a composable. If you did and decided not to, a follow-up is fine as long as it doesn’t get delayed too long—the sooner, the easier for anyone's future work on `KTable`.

radinamatic commented:
Did a couple of passes on this documentation page, haven't noticed any regressions so far, but one point of improvement:  when the focus goes to a column that is **below the sticky one**, there should be a corresponding scroll of the columns in between the sticky ones, so a sighted keyboard user may actually be aware of which column is currently in focus. In the screenshot below, you can see that NVDA properly announces that the focus is on the **Favorite Color** column, which is located **below** the sticky **Options** one, and thus effectively hidden from the sighted keyboard navigation user. The best UX would be to always keep clearly visible whatever/whichever element or part of the page is currently in focus 🙂 

<img width="841" height="563" alt="2025-08-26_05-08-29" src="https://github.com/user-attachments/assets/9eb502f4-c7ad-4214-a63f-9c50168cb11a" />

marcellamaki committed:
Introduce debounce to add shadow on resize

marcellamaki committed:
add scroll into view for tab navigation

marcellamaki committed:
fix linting errors

radinamatic commented:
My previously reported issue is fixed: non-fixed columns are now properly scrolling and fully visible as they are being navigated by keyboard! 👏🏽 👏🏽 👏🏽 

One thing I did notice in the [documentation page](https://deploy-preview-1056--kolibri-design-system.netlify.app/ktable/) is that some content is repeated... Was that intentional? 🤔 

<img width="586" height="2183" alt="repeated" src="https://github.com/user-attachments/assets/b036fad4-6140-46e4-8e76-d480bcf47056" />

radinamatic reviewed (approved):
LGTM! 💯 👏🏽 :shipit:

marcellamaki commented:
Ah, no that looks to be an error with the documentation being duplicated that I missed. I'll update that

AlexVelezLl reviewed (approved):
Tanks @marcellamaki! The visual tests for the new KTable use cases can also be placed in the current `KTableVisualTest` component, but just opened a follow up to track this! https://github.com/learningequality/kolibri-design-system/issues/1127
